src/libraries/CalldataDecoder.sol

```diff
@@ -237,38 +221,69 @@ library CalldataDecoder {
         }
     }
 
-    /// @notice Decode the `_arg`-th element in `_bytes` as a dynamic array
-    /// @dev The decoding of `length` and `offset` is universal,
-    /// whereas the type declaration of `res` instructs the compiler how to read it.
-    /// @param _bytes The input bytes string to slice
+    /// @notice Decode the `_arg`-th element in `_bytes` as `bytes`
+    /// @param _bytes The input bytes string to extract a bytes string from
     /// @param _arg The index of the argument to extract
-    /// @return length Length of the array
-    /// @return offset Pointer to the data part of the array
-    function toLengthOffset(bytes calldata _bytes, uint256 _arg)
-        internal
-        pure
-        returns (uint256 length, uint256 offset)
-    {
+    function toBytes(bytes calldata _bytes, uint256 _arg) internal pure returns (bytes calldata res) {
+        uint256 length;
         uint256 relativeOffset;
         assembly ("memory-safe") {
             // The offset of the `_arg`-th element is `32 * arg`, which stores the offset of the length pointer.
             // shl(5, x) is equivalent to mul(32, x)
             let lengthPtr := add(_bytes.offset, calldataload(add(_bytes.offset, shl(5, _arg))))
+            // the number of bytes in the bytes string
             length := calldataload(lengthPtr)
-            offset := add(lengthPtr, 0x20)
+            // the offset where the bytes string begins
+            let offset := add(lengthPtr, 0x20)
+            // the relative offset of the array from the calldata parameter
             relativeOffset := sub(offset, _bytes.offset)
+            // assign the return parameters
+            res.length := length
+            res.offset := offset
         }
         if (_bytes.length < length + relativeOffset) revert SliceOutOfBounds();
     }
 
     /// @notice Decode the `_arg`-th element in `_bytes` as `bytes`
-    /// @param _bytes The input bytes string to extract a bytes string from
+    /// @param _bytes The input bytes string to extract a bytes array from
     /// @param _arg The index of the argument to extract
-    function toBytes(bytes calldata _bytes, uint256 _arg) internal pure returns (bytes calldata res) {
-        (uint256 length, uint256 offset) = toLengthOffset(_bytes, _arg);
+    function toBytesArray(bytes calldata _bytes, uint256 _arg) internal pure returns (bytes[] calldata res) {
+        uint256 encodingLength;
+        bool isValidEncoding = true;
         assembly ("memory-safe") {
-            res.length := length
-            res.offset := offset
+            let bytesOffset := and(_bytes.offset, OFFSET_OR_LENGTH_MASK)
```

Comment:
Masking done here but not in `toBytes`--intentional?

Reply:
have pushed an update for that https://github.com/Uniswap/v4-periphery/pull/316/commits/2fd239fc0b585a3809e088f78c82ded827abe80e